File: ProjectEuler/p78.cc
// 1, 2, 3, 5, 7, 11, 15, 22, 30, 42
// Partition numbers!
// it's a very slow version
// see p78.py for better 'fast' function
#include <cstdio>
#include <cstring>

const int N = 100;
typedef long long ll;
ll cache[N][N];

ll solve( int n, int m ) {
	if( n == 0 ) return 1;

	ll &ret = cache[n][m];
	if( ret != -1 ) return ret;
	ret = 0;
	for( int i = m ; i >= 0 ; --i ) {
		int remain = n-i;
		if( remain < 0 ) continue;
		ret += solve( remain, i );	
	}

}

int main() {
	memset( cache, -1, sizeof cache );
	for( int i = 1 ; i <= 10 ; ++i ) {
		ll ret = solve( i, i );
		printf("%d %d\n", i, ret-1 );
	}
}

File: ProjectEuler/p78.py
cache = {}

def solve( n, m ):
	if n == 0:
		return 1
	state = (n,m)

	if cache.has_key(state):
		return cache[state]
	cache[state] = 0
	for i in xrange(m,-1,-1):
		remain = n - i
		if remain < 0:
			break
		cache[state] += solve( remain, i )
	return cache[state]

def A000041(n):
	if n == 0:
		return 1
	if cache.has_key(n):
		return cache[n]
	S = 0
	J = n-1
	k = 2
	while 0 <= J:
		T = A000041(J)
		if k/2 % 2 == 1:
			S = S + T
		else:
			S = S - T
		if k % 2 == 1:
			J -= k
		else:
			J -= k / 2
		k += 1
	cache[n] = S
	return cache[n]


for i in xrange(1,100001):
	# ret = solve(i,i)
	ret = A000041(i)
	print i, ret
	if ret % 10**6 == 0:
		print i, ret
		break
